Bug Fix: device of DCK.IDX and DCK.lim not updated

iloveoreoxd opened the pull request:
Fixes this error:

[...]/python3.12/site-packages/DCLS/construct/modules.py", line 457, in forward_vgauss
    P = P + self.lim // 2
        ~~^~~~~~~~~~~~~~~
RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu!

iloveoreoxd committed:
Bug Fix: device of DCK.IDX and DCK.lim not updated

Fixes this error:

[...]/python3.12/site-packages/DCLS/construct/modules.py", line 457, in forward_vgauss
    P = P + self.lim // 2
        ~~^~~~~~~~~~~~~~~
RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu!

iloveoreoxd commented:
This PR fixes an issue where DCK.IDX and DCK.lim remain initialized on CPU after sanity checking. The fix ensures they are moved to the correct device. I used cuda with torch's native Conv2d and Dcls2d.

K-H-Ismail reviewed (changes requested):
Hello @iloveoreoxd,

Thanks for your help. This part of the code was written before I figured out that in fact `register_buffer` function is all we needed there, as in https://github.com/K-H-Ismail/Dilated-Convolution-with-Learnable-Spacings-PyTorch/blob/83de5624a8fa1631ed6e640e1ad1c1796824d249/DCLS/construct/modules.py#L1324

I just didn't make the change for all the other modules `ConstructKernel1d`, ... Could you please move the code from `__init_tmp_variables__` and use `register_buffer` as in  https://github.com/K-H-Ismail/Dilated-Convolution-with-Learnable-Spacings-PyTorch/blob/83de5624a8fa1631ed6e640e1ad1c1796824d249/DCLS/construct/modules.py#L1324 and resubmit the PR?

Thanks.

iloveoreoxd committed:
Use register_buffer function for IDX and lim

This is a Bug Fix for an error where IDX and lim were registered on the wrong device, if the device is changed afterwards:

[...]/python3.12/site-packages/DCLS/construct/modules.py", line 457, in forward_vgauss
P = P + self.lim // 2
~~^~~~~~~~~~~~~~~
RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu!

iloveoreoxd commented:
Thanks for your review! I made the changes and verified that it works with Dcls2d / ConstructKernel2d.

K-H-Ismail committed:
Replace Parameter with tensor for intermediary kernel indices

K-H-Ismail reviewed (approved):
Lgtm. I've also replaced Parameters with `requires_grad=False` in the indices variables by tensors. Thanks!

iloveoreoxd commented:
Perfect!
